src/config.rs

```diff
@@ -20,50 +18,62 @@ pub enum ToolSpec {
         gitlab: String,
         version: VersionReq,
     },
+    Artifactory {
+        artifactory: String,
+        path: String,
+        version: VersionReq,
+    },
 }
 
 impl ToolSpec {
     pub fn cache_key(&self) -> CiString {
         match self {
             ToolSpec::Github { github, .. } => CiString(github.clone()),
             ToolSpec::Gitlab { gitlab, .. } => CiString(format!("gitlab@{}", gitlab)),
+            ToolSpec::Artifactory { artifactory, .. } => CiString(artifactory.clone()),
         }
     }
 
-    pub fn source(&self) -> &str {
+    pub fn source(&self) -> String {
         match self {
             ToolSpec::Github { github: source, .. } | ToolSpec::Gitlab { gitlab: source, .. } => {
-                source
+                source.to_owned()
+            }
+            ToolSpec::Artifactory {
+                artifactory: source,
+                path,
+                ..
+            } => {
+                format!("{}/{}", source, path)
```

Comment:
Any rusty way to avoid making a new String? Maybe add a new field and move line 47 to when we initialize the ToolSpec::Artifactory?

Reply:
Hmm, I see your point; our definition of source gets more complicated with the artifactory definition, I almost wonder if we could separate source and path in the abstraction in general?

That way you'd have sources like: `https://github.com`, `https://gitlab.com`, `https://somehost.artifactory.com`, `https://otherhost.artifactory.com`, etc.

And then paths like: `Roblox/foreman`, `path/to/artifact`, etc.

Reply:
An alternative would be to have every tool spec be composed of `source` (more like host) + `path` + `version`, in which case we'd maybe want the existing stuff like `rotrieve = { github = "roblox/rotriever", version = "0.5.12" }` is just a shorthand for:
```toml
[tools.rotrieve]
source = "https://github.com"
path = "roblox/rotriever"
version = "0.5.12"
```